File: trees_and_graphs/binary_tree_right_view.py
# binary tree right side view
from collections import deque

class Node:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None

def right_view(root):

    if not root:
        return []
    
    queue = deque([root])
    result = []

    while queue:
        level_size = len(queue)
        rightmost = queue[level_size - 1]
        result.append(rightmost.val)
        
        for _ in range(level_size):
            node = queue.popleft()

            if node.left:
                queue.append(node.left)

            if node.right:
                queue.append(node.right)

    return result        

# tests
n = Node(3)
n.left = Node(4)
n.right = Node(4)
n.left.left = Node(7)
n.left.right = Node(1)
n.right.left = Node(1)
n.right.right = Node(7)
assert right_view(n) == [3,4,7]

n = Node(3)
assert right_view(n) == [3]

n = Node(3)
n.left = Node(4)
n.right = Node(4)
n.left.left = Node(7)
n.left.right = Node(1)
n.right.left = Node(1)
n.right.right = Node(7)
n.left.right.left = Node(0)

assert right_view(n) == [3,4,7,0]

print("all tests passed.")

File: trees_and_graphs/binary_tree_right_view.php
<?php

class Node {
    public function __construct($val){
        $this->val = $val;
        $this->left = null;
        $this->right = null;
    }
}
class Solution 
{
    public function right_view(Node $node) : array {

        $queue = [$node];

        $result = [];

        while ($queue) {
            $level_size = count($queue);

            for($i = 0; $i < $level_size; $i++) {
                $node = array_shift($queue);
                if ($i == $level_size - 1) {
                    $result[] = $node->val;
                }

                if (isset($node->left)) {
                    $queue[] = $node->left;
                }

                if (isset($node->right)) {
                    $queue[] = $node->right;
                }
            }
        }

        return $result;
    }
}


$node = new Node(5);
$node->left = new Node(2);
$node->right = new Node(4);
$node->left->left = new Node(7);

$s = new Solution();
$right_view = $s->right_view($node);

echo implode("\n",$right_view);
echo "\n";

?>